src/Storages/IMessageProducer.cpp

```diff
@@ -0,0 +1,37 @@
+#include <Storages/IMessageProducer.h>
+#include <Core/BackgroundSchedulePool.h>
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+
+void ConcurrentMessageProducer::start(const ContextPtr & context)
+{
+    initialize();
+    producing_task = context->getSchedulePool().createTask(getProducingTaskName(), [this]
+    {
+        producingTask();
+        /// Notify that producing task is finished.
+        task_finished.store(true);
+        task_finished.notify_all();
+    });
+    producing_task->activateAndSchedule();
+}
+
+void ConcurrentMessageProducer::finish()
+{
+    LOG_DEBUG(&Poco::Logger::get("ConcurrentMessageProducer"), "finish");
+
+    /// We should execute finish logic only once.
+    if (finished.exchange(true))
+        return;
+
+    stopProducingTask();
+    /// Wait until producing task is finished.
+    task_finished.wait(false);
+    producing_task->deactivate();
```

Comment:
AFAIK `producing_task->deactivate(); == (task_finished == true)` as `deactivate` waits for task to finish its execution, and the task sets task_finished = true right before finishing, so we can remove `task_finished.wait(false);` from here?

Reply:
and as I see, looks like it is the only place where this flag is used - but if deactivate waits for task to be finihsed, then it is not needed at all.

Reply:
Didn't know that `deactivate` is waiting for task to finish. Thanks! I removed `task_finished` completely.